### PR TITLE
[NG] Fix for focus issue when escaping forms in modals

### LINKED ISSUE #1194: Modal and Wizard with form throws an error when pressing escape to close
**Select one ...**  (check one with "x")
```
[X] bug
[ ] feature request
[ ] enhancement
```

### Expected behavior

Don't throw an error when pushing "escape" to close a modal with a form.

### Actual behavior

1) Open the modal
2) Click the input (so it's `ng-touched`)
3) Press "escape"
4) Observe the error thrown in the console

<details>
<summary>Error</summary>
<pre>
ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'true'. Current value: 'false'.
    at viewDebugError (errors.ts:30)
    at expressionChangedAfterItHasBeenCheckedError (errors.ts:29)
    at checkBindingNoChanges (util.ts:145)
    at checkNoChangesNodeInline (view.ts:477)
    at checkNoChangesNode (view.ts:568)
    at debugCheckNoChangesNode (services.ts:557)
    at debugCheckRenderNodeFn (services.ts:500)
    at Object.eval [as updateRenderer] (app.component.html:17)
    at Object.debugUpdateRenderer [as updateRenderer] (services.ts:477)
    at checkNoChangesView (view.ts:383)
</pre>
</details>

### Reproduction of behavior
https://plnkr.co/edit/FJLwhI13nzeETVqWBYI2?p=preview

### Environment details
* **Angular version:** 2.0.X

* **Clarity version:** 0.9.7

* **OS and version:** Mac

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

### PR DESCRIPTION
**Pull request by hippee-lee**
- Added a reference to the previous active element
- Adds a method to let focus trap users set focus in the previousActiveElement if they need to
- closes #1194

Signed-off-by: Matt Hippely <mhippely@vmware.com>

### CODE REVIEW

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.ts [hunk 1]
```diff
@@ -19,7 +22,16 @@ export class FocusTrapDirective implements AfterViewInit {
     }
 
     ngAfterViewInit() {
+        this._previousActiveElement = <HTMLElement>document.activeElement;
```
Comment (youdz) [inline src/clarity-angular/utils/focus-trap/focus-trap.directive.ts]: You should probably cast it to `Document`, it's an [actual interface](https://developer.mozilla.org/en-US/docs/Web/API/Document).

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.ts [hunk 2]
```diff
@@ -19,7 +22,16 @@ export class FocusTrapDirective implements AfterViewInit {
...
+        // console.log("Focus trap directive view init, activeElement: ", document.activeElement);
+        // console.log("Focus trap directive _previousActiveElement: ", this._previousActiveElement);
```
Comment (youdz) [inline src/clarity-angular/utils/focus-trap/focus-trap.directive.ts]: A couple commented out logs to clean.

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.ts [hunk 3]
```diff
@@ -19,7 +22,16 @@ export class FocusTrapDirective implements AfterViewInit {
...
         const nativeElement: HTMLElement = this.elementRef.nativeElement;
         nativeElement.setAttribute("tabindex", "0");
     }
+
+    public setPreviousFocus(): void {
```
Comment (youdz) [inline src/clarity-angular/utils/focus-trap/focus-trap.directive.ts]: Can you please add:
```typescript
ngOnDestroy() {
    this.setPreviousFocus();
}
```

It doesn't cost much and adds safety for other focus traps in the future that are not modals.

File: src/app/modal/modal-form.demo.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    model = new Hero(18, "Dr IQ", this.powers[0], "Chuck Overstreet");
+
+    submitted = false;
+
+    onSubmit(form: any) {
+        this.submitted = true;
+        console.log(form);
+    }
+
+    // TODO: Remove this when we're done
```
Comment (gnomeontherun) [inline src/app/modal/modal-form.demo.ts]: Are we done so this should be removed?

File: src/app/modal/modal-form.demo.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    basic: boolean = false;
+    powers = ["Really Smart", "Super Flexible", "Super Hot", "Weather Changer"];
+
+    model = new Hero(18, "Dr IQ", this.powers[0], "Chuck Overstreet");
+
+    submitted = false;
+
+    onSubmit(form: any) {
+        this.submitted = true;
+        console.log(form);
```
Comment (gnomeontherun) [inline src/app/modal/modal-form.demo.ts]: log

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.ts (hunk 3 above)
Comment (hippee-lee) [reply src/clarity-angular/utils/focus-trap/focus-trap.directive.ts]: Will add.

File: src/app/modal/modal-form.demo.ts (hunk 4 above)
Comment (hippee-lee) [reply src/app/modal/modal-form.demo.ts]: That was from the demo code I copied in to develop with. Will remove the comment.

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.ts [hunk 6]
```diff
@@ -3,11 +3,14 @@
...
-import {AfterViewInit, Directive, ElementRef, HostListener} from "@angular/core";
+import {DOCUMENT} from "@angular/common";
+import {AfterViewInit, Directive, ElementRef, HostListener, Inject} from "@angular/core";
 
 @Directive({selector: "[clrFocusTrap]"})
 export class FocusTrapDirective implements AfterViewInit {
-    constructor(public elementRef: ElementRef) {}
+    private _previousActiveElement: HTMLElement;
+
+    constructor(public elementRef: ElementRef, @Inject(DOCUMENT) document: any) {}
```
[the remark this answers is not in the document]
Comment (hippee-lee) [reply src/clarity-angular/utils/focus-trap/focus-trap.directive.ts]: Will add the correct type.

File: src/app/modal/modal-form.demo.ts (hunk 5 above)
Comment (hippee-lee) [reply src/app/modal/modal-form.demo.ts]: Will remove.

File: src/clarity-angular/utils/focus-trap/focus-trap.directive.ts (hunk 2 above)
Comment (hippee-lee) [reply src/clarity-angular/utils/focus-trap/focus-trap.directive.ts]: Will remove.

File: package.json [hunk 7]
```diff
@@ -16,9 +16,6 @@
     "@angular/upgrade": "^4.3.0",
     "@webcomponents/custom-elements": "^1.0.0",
     "angular": "1.5.x",
-    "clarity-angular": "file:dist/npm/clarity-angular",
```
Comment (gnomeontherun) [inline package.json]: Thank you

**Comment by hippee-lee** [comment]
Added do-not-merge b/c I need a second round of reviews for the feedback changes.

**Comment by gnomeontherun** [comment]
Looks good from my end.

**Pull request merged**